[조이] 뉴스 스탠드 3주차 - Final

youzysu opened the pull request:
# DEMO

# 주요 작업

youzysu committed:
refactor: 코드 리뷰 반영

- element private에서 public readonly로 수정

youzysu committed:
feat: 언론사 기사 영역 상단 언론사 정보 컴포넌트 생성

- 하위 컴포넌트 분리 및 모듈화

youzysu committed:
feat #27: 리스트 보기 언론사 기사 영역 UI

youzysu committed:
feat: 리스트 보기 언론사 순서 랜덤으로 정하기

- 리스트 보기 데이터 가져올 때 기사 리스트만 랜덤으로 섞는다.

youzysu commented:
## 페어리뷰 때 논의하고 싶은 내용
### 1. [스토어 로직 개선 방향](https://github.com/youzysu/fe-max--newsstand/issues/21)
### 2. 변경된 부분만 반영하여 렌더링시키는 방법 
- Current: 그 컴포넌트의 현재 상태를 저장하여 새로운 props와 비교하기 위해 state라는 property에 저장해두고 있다. reducer에서 새로운 객체를 생성하기 때문에 `!==` 로 객체를 비교하여 새로운 값이면 반영
- 변경을 감지해야 하는 상태가 하나라도 변경되면 하위 컴포넌트가 렌더링되도록 하는 구조
- 고민
  1. State 네이밍이 부적절한가?
  2. 그렇다면 prevProps?
  3. 다른 방법은? 웹 컴포넌트에서 oldVal, newVal 반영하듯이?
### 3. Store에서 상태로 관리하는 기준
- 여러 다른 로직에서 반복되는 작업인 경우에 그래도 어떻게 반복하지 않고 구할 수 없을까? 마치 상수처럼 별도로 관리하는건 어떨까? (ex. 카테고리 언론사의 개수를 상단의 fieldTab에서 보여주고, 버튼 클릭 시에도 다음 index를 결정하기 위해 필요함)
> Think of state as the minimal set of changing data that your app needs to remember. 
Figure out the absolute minimal representation of the state your application needs and compute everything else on-demand. 
For example, if you’re building a shopping list, you can store the items as an array in state. If you want to also display the number of items in the list, **don’t store the number of items as another state value—instead, read the length of your array.** - [Thinking in React](https://react.dev/learn/thinking-in-react)

youzysu committed:
feat #27: 리스트보기 좌우 화살표 클릭 시 언론사 기사 영역 이동하기

- 이전/다음 언론사로 이동한다.
- 카테고리 마지막 언론사인 경우 다음 카테고리로 이동한다.
- 카테고리의 첫 언론사인 경우 이전 카테고리로 이동한다.

hjsong333 commented on src/components/MediaViewer/ListViewer/PressArticle/MainArticle.ts:
```diff
@@ -0,0 +1,24 @@
+  private thumbnail = createElement('IMG', { class: styles.thumbnail });
+  private title = createElement('A', { class: `body-md ${styles.mainTitle}` });
+
+  constructor() {
+    this.element.append(this.thumbnail, this.title);
+  }
+
+  public render({ thumbnail, mainArticle }: MainArticleProps) {
+    this.thumbnail.setAttribute('src', thumbnail.img);
+    this.thumbnail.setAttribute('alt', thumbnail.title);
+    this.title.setAttribute('href', mainArticle.link);
+    this.title.textContent = mainArticle.title;
+  }
```
생성 타이밍에 element 를 append 해두고 props를 받아서 attribute와 textContent를 채워주는 방식 깔끔하네요 👍

hjsong333 commented on src/components/MediaViewer/ListViewer/PressArticle/PressArticle.ts:
```diff
@@ -0,0 +1,39 @@
+}
+
+export default class PressArticle {
+  public readonly element = createElement('DIV', { class: styles.pressArticle });
+  private pressInfo = new PressInfo();
+  private article = createElement('ARTICLE', { class: styles.article });
+  private mainArticle = new MainArticle();
+  private subArticle = new SubArticle();
```
이런 방식으로 초기화하신 이유가 있을까요?
생성자 내에서 초기화하는 것과 차이가 있는지 궁금하네요.

hjsong333 commented on src/components/MediaViewer/ListViewer/PressArticle/SubArticle.ts:
```diff
@@ -0,0 +1,33 @@
+      const articleLink = this.articleLinks[index];
+      articleLink.setAttribute('href', link);
+      articleLink.textContent = title;
+      articleItem.append(articleLink);
```
articleItem을 생성자에서 append 해주고 여기서는 attribute, textContent만 변경해줘도 내용이 변경되지 않을까요?

youzysu replied:
넵! 통일성을 위해 아래처럼 수정했어요! 감사합니다 👍 

```ts
  constructor() {
    this.articleList.append(...this.articleListItems);
    this.articleListItems.forEach((articleItem, index) => articleItem.append(this.articleLinks[index]));
    this.element.append(this.articleList, this.caption);
  }

  public render({ pressName, subArticleList }: SubArticleProps) {
    this.articleLinks.forEach((articleLink, index) => {
      const { title, link } = subArticleList[index];
      articleLink.setAttribute('href', link);
      articleLink.textContent = title;
    });
    this.caption.textContent = `${pressName} 언론사에서 직접 편집한 뉴스입니다.`;
  }
```

youzysu committed:
fix: 리스트 보기 언론사 인덱스 변경하기

- refactor 필요: index 관리하는 방식을 다르게 하거나, 데이터 요청 단위를 쪼개기
- 카테고리의 마지막 언론사의 다음에는 다음 카테고리의 첫 언론사로 넘어간다.
- 단, 마지막 카테고리의 마지막 언론사 다음에는 첫 카테고리의 첫 언론사로 돌아온다.
- 카테고리의 첫 언론사의 이전에는 이전 카테고리의 마지막 언론사로 넘어간다.
- 단, 첫 카테고리의 첫 언론사 이전에는 마지막 카테고리의 마지막 언론사로 넘어간다.

youzysu committed:
feat: 카테고리를 클릭하면 해당 카테고리로 바로 이동하기

youzysu committed:
feat: 현재 카테고리 탭 배경 프로그레스바 애니메이션 적용

- 프로그레스 바 요소 및 카테고리 이름과 개수 z-index 추가

youzysu committed:
refactor: 코드 리뷰 반영

- 생성자에서 요소 붙여주기

youzysu committed:
feat: 한 언론사 당 20초 동안 화면에 보여지고, 20초가 지나면 다음 언론사로 이동하기

- 프로그레스바 animation이 끝날 때 리스트 이동하는 액션을 실행한다.

youzysu committed:
feat: 선택된 카테고리가 아닌 다른 카테고리에 마우스 호버 시 밑줄이 생긴다.

youzysu committed:
feat: 메인 아티클에 마우스 호버 시 썸네일 이미지 확대 및 뉴스 타이틀 밑줄 효과

youzysu committed:
feat: 리스트 보기에서 [구독하기] 버튼 클릭 시 스낵바 보여주기

youzysu committed:
fix: 리스트보기에서 해지하기 버튼 클릭 시에는 스낵바가 보여지지 않기

- .subscribed 클래스 포함 여부로 버튼을 구분한다.

youzysu committed:
refactor: 리스트 보기 다음/이전 언론사 로직 별도 함수로 분리

- getNextCategoryPress, getPrevCategoryPress
